tendrl/gluster_integration/sds_sync/__init__.py

```diff
@@ -930,3 +924,91 @@ def get_volume_alert_counts():
+        if not err:
+            out = out.encode('utf8')
+            devlist = map(
+                lambda line: dict(zip(keys, line.split(' '))),
+                out.splitlines()
+            )
+            block_devices = {}
+            for dev_info in devlist:
+                if dev_info["TYPE"] == "disk":
+                    # create new block device list
+                    block_devices = {dev_info["KNAME"]: dev_info}
+                else:
+                    block_devices[dev_info["KNAME"]] = dev_info
+                # find ancestors if mountpoint present
+                if dev_info["MOUNTPOINT"]:
+                    ancestor_details = find_ancestors(
+                        block_devices, dev_info["KNAME"]
+                    )
+                    if dev_info["MOUNTPOINT"] in devicetree.keys():
+                        # if different block device mounted in same mountpoint
+                        # then append the disk and partition information under
+                        # same mountpoint
+                        disks = devicetree[dev_info["MOUNTPOINT"]]["disks"]
+                        devicetree[dev_info["MOUNTPOINT"]]["disks"] = list(
+                            set(disks + ancestor_details["disks"])
+                        )
+                        devicetree[dev_info["MOUNTPOINT"]]["partitions"] = \
+                            list(
+                                set(ancestor_details["partitions"])
+                            )
+                    else:
+                        devicetree[dev_info["MOUNTPOINT"]] = {}
+                        devicetree[dev_info["MOUNTPOINT"]]["disks"] = \
+                            ancestor_details["disks"]
+                        devicetree[dev_info["MOUNTPOINT"]]["partitions"] = \
+                            ancestor_details["partitions"]
+                        devicetree[dev_info["MOUNTPOINT"]]["device_info"] = {}
+                        devicetree[dev_info["MOUNTPOINT"]]["device_info"][
+                            "size"] = dev_info["SIZE"]
+                        devicetree[dev_info["MOUNTPOINT"]]["device_info"][
+                            "mountpoint"] = dev_info["MOUNTPOINT"]
+                        devicetree[dev_info["MOUNTPOINT"]]["device_info"][
+                            "type"] = dev_info["TYPE"]
+                        devicetree[dev_info["MOUNTPOINT"]]["device_info"][
+                            "size"] = dev_info["SIZE"]
+        return devicetree
+    except Exception as ex:
+        logger.log(
+            "error",
+            NS.publisher_id,
+            {"message": "Unable to generate devicetree.err: %s" % ex}
+        )
+
+
+def find_ancestors(block_devices, device_name):
+    # find ancestors including current dev from bottom to top
+    ancestor_details = {}
+    ancestor_details["disks"] = []
+    ancestor_details["partitions"] = []
+    while True:
+        if device_name in block_devices.keys():
```

Comment:
So if device not found in block_devices, we eventually return a blank ancestors details. Right?

Reply:
if any mount point present then only we are calling a function to find its an ancestor, so definitely ancestor will present.

if any disk mounted without partition  like:
vda                                 252:0    0   41G  0 disk  /boot

also, consider itself ancestor. 

onemore example: 
vda                                 252:0    0   41G  0 disk  
├─vda1                              252:2    0    1G  0 part /boot

this will give a result like {"/boot": "disks": [vda], "partitions": [vda1]}
for vda1 it will cosider itself also ancestor and added into partitions.

Reply:
Ack